### ISSUE TITLE
building on arm64 fails with commit 523e51

### ISSUE THREAD

**Issue by omajid** [dotnet/source-build#1411]
With commit `523e51da8431653b49e8fee3c1cde0c96df43c5d` (the latest `release/3.1` branch), I can no longer build on arm64. A simple `./build.sh` fails. 

I can not reproduce the failure with `d36f3bea704efd4194aebca03acc0e5f216482f7`, the previous commit in the `release/3.1` branch.

from the build log:

```
  [SourceBuiltSdkResolver] Overriding Microsoft.DotNet.Arcade.Sdk with 'ARCADE'
Restore completed in 3.03 sec for /home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj.
Restore completed in 3.3 sec for /home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/core-sdk-tasks/core-sdk-tasks.csproj.                                                                                Restore completed in 5.84 sec for /home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/redist/redist.csproj.
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error : The type initializer for 'LibGit2Sharp.Core.NativeMethods' threw an exception. [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at LibGit2Sharp.Core.NativeMethods.git_buf_free(GitBuf buf) [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at LibGit2Sharp.Core.Proxy.git_buf_free(GitBuf buf) [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at LibGit2Sharp.Core.Handles.GitBuf.Dispose() [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at LibGit2Sharp.Core.Proxy.ConvertPath(Func`2 pathRetriever) [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at LibGit2Sharp.Core.Proxy.git_repository_discover(FilePath start_path) [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at LibGit2Sharp.Repository.Discover(String startingPath) [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]                                                                                                                
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at Microsoft.Build.Tasks.Git.GitOperations.LocateRepository(String directory) in /_/src/Microsoft.Build.Tasks.Git.Operations/GitOperations.cs:line 26 [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]                   
/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets(20,5): error :    at Microsoft.Build.Tasks.Git.RepositoryTasks.LocateRepository(LocateRepository task) in  /_/sr /Microsoft.Build.Tasks.Git.Operations/RepositoryTasks.cs:line 58 [/home/omajid/source-build/bin/src/core-sdk.cd82f021f4c2c8901bd7975fcaefc34fdeec3ded/src/Microsoft.Dotnet.Sdk.Internal/Microsoft.Dotnet.Sdk.Internal.csproj]
  ##vso[task.logissue type=error;sourcepath=/home/omajid/source-build/packages/restored/microsoft.build.tasks.git/1.0.0-beta2-18618-05/build/Microsoft.Build.Tasks.Git.targets;linenumber=20;columnnumber=5;code=;]The type initializer for 'LibGit2Sharp.Core.NativeMethods' threw an exception.%0A   at LibGit2Sharp.Core.NativeMethods.git_buf_free(GitBuf buf)%0A   at LibGit2Sharp.Core.Proxy.git_buf_free(GitBuf buf)%0A   at LibGit2Sharp.Core.Handles.GitBuf.Dispose()%0A   at LibGit2Sharp.Core.Proxy.ConvertPath(Func`2 pathRetriever)%0A   at LibGit2Sharp.Core.Proxy.git_repository_discover(FilePath start_path)%0A   at LibGit2Sharp.Repository.Discover(String startingPath)%0A   at Microsoft.Build.Tasks.Git.GitOperations.LocateRepository(String directory) in /_/src/Microsoft.Build.Tasks.Git.Operations/GitOperations.cs:line 26%0A   at Microsoft.Build.Tasks.Git.RepositoryTasks.LocateRepository(LocateRepository task) in /_/src/Microsoft.Build.Tasks.Git.Operations/RepositoryTasks.cs:line 58
```
It seems like ~~arcade~~ libgit2sharp is broken on arm64 (again) but not sure why/how. Trying to revert the versions in `global.json` and `eng/Version.Details.xml` doesn't work and leads to errors about missing frameworks.

Logs:

[core-sdk.log](https://github.com/dotnet/source-build/files/3908678/core-sdk.log)
[core-sdk.binlog.tar.gz](https://github.com/dotnet/source-build/files/3908679/core-sdk.binlog.tar.gz)

**Comment by omajid** [comment]
cc @tmds @RheaAyase

**Comment by tmds** [comment]
> microsoft.build.tasks.git

This package has a dependency on libgit2sharp `0.26.0-preview-0070` which we know doesn't  work well on arm64 (https://github.com/dotnet/sourcelink/blob/master/eng/Versions.props#L25).

I guess we didn't use this code path previously. This may be related to some source-link functionality that the build is trying to do. microsoft.build.tasks.git is part of https://github.com/dotnet/sourcelink.

**Comment by omajid** [comment]
We build `sourcelink` (commit `855afb5af9ea79e63dd9c051a7cfe0941bbb5ced`) as part of `source-build`'s `release/3.1` branch. Would it be possible to use that version (including any arm64 fixes required) when building `core-sdk` here?

**Comment by omajid** [comment]
This following change makes things work, but I dont understand why. And I certainly dont know if it's the right fix.

```diff
diff --git a/patches/applicationinsights-dotnet/0001.patch b/patches/applicationinsights-dotnet/0001.patch
new file mode 100644
index 0000000..8aead70
--- /dev/null
+++ b/patches/applicationinsights-dotnet/0001.patch
@@ -0,0 +1,11 @@
+--- a/LOGGING/src/ILogger/ILogger.csproj
++++ b/LOGGING/src/ILogger/ILogger.csproj
+@@ -54,7 +54,7 @@
+ 
+   <ItemGroup>
+     <ProjectReference Include="..\..\..\BASE\src\Microsoft.ApplicationInsights\Microsoft.ApplicationInsights.csproj" />
+-    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-18618-05">
++    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-19367-01">
+       <PrivateAssets>all</PrivateAssets>
+       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
+     </PackageReference>
diff --git a/patches/core-sdk/0003-Enable-building-on-arm64-machines.patch b/patches/core-sdk/0003-Enable-building-on-arm64-machines.patch
index 51937e7..d222c34 100644
--- a/patches/core-sdk/0003-Enable-building-on-arm64-machines.patch
+++ b/patches/core-sdk/0003-Enable-building-on-arm64-machines.patch
@@ -46,6 +46,18 @@ index b65a72410..be3834859 100644
    <PropertyGroup>
      <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
      <DebugType>embedded</DebugType>
+diff --git a/eng/Versions.props b/eng/Versions.props
+--- a/eng/Versions.props
++++ b/eng/Versions.props
+@@ -117,7 +117,7 @@
+     <VersionToolsVersion>$(BuildTasksFeedToolVersion)</VersionToolsVersion>
+     <DotnetDebToolVersion>2.0.0</DotnetDebToolVersion>
+     <MicrosoftNETTestSdkVersion>15.8.0</MicrosoftNETTestSdkVersion>
+-    <MicrosoftSourceLinkVersion>1.0.0-beta2-18618-05</MicrosoftSourceLinkVersion>
++    <MicrosoftSourceLinkVersion>1.0.0-beta2-19367-01</MicrosoftSourceLinkVersion>
+   </PropertyGroup>
+   <PropertyGroup>
+     <!-- pinned dependency. This package is not being produced outside of the 2.0 branch of corefx and should not change. -->
 diff --git a/src/redist/targets/Crossgen.targets b/src/redist/targets/Crossgen.targets
 index 8d3091307..931dff2d9 100644
 --- a/src/redist/targets/Crossgen.targets
diff --git a/support/additional-prebuilts-to-delete.txt b/support/additional-prebuilts-to-delete.txt
index 04e340c..302fe44 100644
--- a/support/additional-prebuilts-to-delete.txt
+++ b/support/additional-prebuilts-to-delete.txt
@@ -7,7 +7,7 @@ microsoft.aspnetcore.analyzers.3.0.0-rc2.19465.2.nupkg
 microsoft.aspnetcore.developercertificates.xplat.3.0.0-rc2.19465.2.nupkg
 microsoft.aspnetcore.mvc.analyzers.3.0.0-rc2.19465.2.nupkg
 microsoft.aspnetcore.mvc.api.analyzers.3.0.0-rc2.19465.2.nupkg
-microsoft.build.tasks.git.1.0.0-beta2-18618-05.nupkg
+microsoft.build.tasks.git.1.0.0-beta2-19367-01.nupkg
 microsoft.codeanalysis.analyzers.2.6.1.nupkg
 microsoft.codeanalysis.common.2.9.0.nupkg
 microsoft.codeanalysis.csharp.2.9.0.nupkg
@@ -17,9 +17,9 @@ microsoft.dotnet.genapi.*.nupkg
 microsoft.dotnet.signtool.*.nupkg
 microsoft.net.compilers.toolset.*.nupkg
 microsoft.netcore.app.host.linux-x64.3.0.0.nupkg
-microsoft.sourcelink.common.1.0.0-beta2-18618-05.nupkg
-microsoft.sourcelink.github.1.0.0-beta2-18618-05.nupkg
-microsoft.sourcelink.vsts.git.1.0.0-beta2-18618-05.nupkg
+microsoft.sourcelink.common.1.0.0-beta2-19367-01.nupkg
+microsoft.sourcelink.github.1.0.0-beta2-19367-01.nupkg
+microsoft.sourcelink.vsts.git.1.0.0-beta2-19367-01.nupkg
 microsoft.visualstudio.setup.configuration.interop.1.16.30.nupkg
 runtime.*.runtime.native.system.security.cryptography.apple.*.nupkg
 runtime.*.runtime.native.system.security.cryptography.openssl.*.nupkg
diff --git a/tools-local/prebuilt-baseline-online.xml b/tools-local/prebuilt-baseline-online.xml
index 8373234..c287043 100644
--- a/tools-local/prebuilt-baseline-online.xml
+++ b/tools-local/prebuilt-baseline-online.xml
@@ -62,7 +62,7 @@
     <Usage Id="Microsoft.Build.Framework" Version="15.7.179" IsDirectDependency="true" />
     <Usage Id="Microsoft.Build.Tasks.Core" Version="15.1.548" IsDirectDependency="true" />
     <Usage Id="Microsoft.Build.Tasks.Core" Version="15.7.179" IsDirectDependency="true" />
-    <Usage Id="Microsoft.Build.Tasks.Git" Version="1.0.0-beta2-18618-05" />
+    <Usage Id="Microsoft.Build.Tasks.Git" Version="1.0.0-beta2-19367-01" />
     <Usage Id="Microsoft.Build.Traversal" Version="2.0.2" />
     <Usage Id="Microsoft.Build.Utilities.Core" Version="15.1.548" />
     <Usage Id="Microsoft.Build.Utilities.Core" Version="15.1.1012" IsDirectDependency="true" />
@@ -135,9 +135,9 @@
     <Usage Id="Microsoft.NETCore.Targets" Version="2.1.0" />
     <Usage Id="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0-preview.2" IsDirectDependency="true" IsAutoReferenced="true" />
     <Usage Id="Microsoft.NETFramework.ReferenceAssemblies.net472" Version="1.0.0-preview.2" />
-    <Usage Id="Microsoft.SourceLink.Common" Version="1.0.0-beta2-18618-05" />
-    <Usage Id="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-18618-05" IsDirectDependency="true" IsAutoReferenced="true" />
-    <Usage Id="Microsoft.SourceLink.Vsts.Git" Version="1.0.0-beta2-18618-05" IsDirectDependency="true" IsAutoReferenced="true" />
+    <Usage Id="Microsoft.SourceLink.Common" Version="1.0.0-beta2-19367-01" />
+    <Usage Id="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-19367-01" IsDirectDependency="true" IsAutoReferenced="true" />
+    <Usage Id="Microsoft.SourceLink.Vsts.Git" Version="1.0.0-beta2-19367-01" IsDirectDependency="true" IsAutoReferenced="true" />
     <Usage Id="Microsoft.TargetingPack.NETFramework.v4.7.2" Version="1.0.0" IsDirectDependency="true" />
     <Usage Id="Microsoft.TargetingPack.Private.WinRT" Version="1.0.5" />
     <Usage Id="Microsoft.VisualStudio.Setup.Configuration.Interop" Version="1.16.30" />
```

**Comment by tmds** [comment]
> +-    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-18618-05">
++    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-19367-01">

The newer version doesn't depend on libgit2sharp.

**Comment by dseefeld** [comment]
Should be fixed by #1415.  Assigning to @crummel to confirm.

**Comment by omajid** [comment]
I retried builds of release/3.1 on arm64. They are now working out of the box for me. Thanks for fixing this!
